### PR TITLE
Serve Feed Content From Elasticsearch

### PR DESCRIPTION
**Pull request by mstruve**
## What type of PR is this? (check all applicable)
**- [x] Feature**

## Description
This PR switches our search for the main feed content from Algolia to Elasticsearch. Couple of changes that I made.
* Comments are searched SEPARATE from Articles. If you search for a phrase that is in a comment you will see the comment returned in the Comment tab but not the Article. 
* I updated our highlighting to return **2** fragments as opposed to the original 1. This allows for a little more insight into what matched in your article
* Podcasts also return highlighted text from the Podcast body so once again you have a little insight into how the Podcast matched your search. 

**WHAT I NEED**
Testing: Please feel free to pull this branch down and run it through the wringer. I have done a lot of manual testing myself but the more people that explore this the more likely we are to find the edge cases. 
Before you test, make sure ES is up to date (run bin/setup)

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/6#card-32289432

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
Podcast Example Return
Screen Shot 2020-03-25 at 1 39 24 PM
Comment Example Return
Screen Shot 2020-03-25 at 1 39 33 PM
Multiple Highlighted Fragments
Screen Shot 2020-03-23 at 9 59 47 AM

## Added tests?
- [x] no, I only updated the ones that already existed.
I am VERY open to getting some help to write some more frontend tests for this. 

## Added to documentation?
- [x] no documentation needed

alt_text

### CODE REVIEW

File: app/assets/javascripts/initializers/initScrolling.js.erb [hunk 1]
```diff
@@ -179,40 +179,46 @@ function fetchNextPodcastPage(el) {
...
+  const searchHash = { ...{ per_page: 15, page: nextPage }, ...JSON.parse(params) };
 
-  var searchObj = {
-      hitsPerPage: 15,
-      page: nextPage,
-      attributesToHighlight: [],
-      tagFilters,
-      filters: (requiresApproval === 'true') ? 'approved:true' : ''
+  if (tag && tag.length > 0) {
+    searchHash["tag_names"] = (searchHash["tag_names"] === undefined) ? [] : searchHash["tag_names"];
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initScrolling.js.erb]: If `searchHash["tag_names"]` is `null` or `undefined`, it will take the value of `[]`

```suggestion
    searchHash["tag_names"] = searchHash["tag_names"] || [];
```

File: app/assets/javascripts/initializers/initScrolling.js.erb [hunk 2]
```diff
@@ -179,40 +179,46 @@ function fetchNextPodcastPage(el) {
...
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then(response => response.json())
+    .then(function searchDone(content) {
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initScrolling.js.erb]: This can be an arrow function.

```suggestion
    .then((content) => {
```

File: app/assets/javascripts/utilities/buildArticleHTML.js.erb [hunk 3]
```diff
@@ -37,42 +37,41 @@ function buildArticleHTML(article) {
...
       flareTag = "<span class='tag-identifier'>podcast</span>"
     }
+    if (article.class_name == "Comment"){
+      flareTag = "<span class='tag-identifier'>comment</span>"
+    }
     if (article.class_name == "User"){
       flareTag = "<span class='tag-identifier' style='background:#5874d9;color:white;'>person</span>"
     }
-    var rc = article.positive_reactions_count || article.reactions_count
+    var rc = article.positive_reactions_count
```
Comment (nickytonline) [inline app/assets/javascripts/utilities/buildArticleHTML.js.erb]: +1 for phasing out `article.reactions_count`. 👏🏻

File: app/assets/javascripts/initializers/initScrolling.js.erb (hunk 1 above)
Comment (mstruve) [reply app/assets/javascripts/initializers/initScrolling.js.erb]: I totally thought I tried this in a console and it didnt work, I just retried it and it does. 🤦‍♀  Thank you!

File: app/views/articles/_search.html.erb [hunk 4]
```diff
@@ -45,68 +43,64 @@
...
+      method: 'GET',
+      headers: {
+        Accept: 'application/json',
+        'X-CSRF-Token': window.csrfToken,
+        'Content-Type': 'application/json',
+      },
+      credentials: 'same-origin',
+    })
+      .then(response => response.json())
       .then(function searchDone(content) {
```
Comment (nickytonline) [inline app/views/articles/_search.html.erb]: This can be an arrow function like up above.

```suggestion
    .then((content) => {
```

File: app/javascript/articles/components/SearchSnippet.jsx [hunk 5]
```diff
@@ -1,45 +1,25 @@
...
-      let startingEllipsis = '';
-
-      if (firstSnippetChar.toLowerCase() !== firstSnippetChar.toUpperCase()) {
-        startingEllipsis = '…';
-      }
-
-      commentsBlobSnippet = `${startingEllipsis +
-        snippetResult.comments_blob.value}… <i>(comments)</i>`;
+      bodyTextSnippet = `${startingEllipsis +
+        highlightText.body_text.join('...')}…`;
```
Comment (nickytonline) [inline app/javascript/articles/components/SearchSnippet.jsx]: Is `join` being used now because the `body_text` begins and ends with `...`?

File: app/views/articles/_search.html.erb [hunk 6]
```diff
@@ -45,68 +43,64 @@
...
-    }
+  function search(query, filters) {
+    var hashtags = query.match(/#\w+/g);
+    var searchTerm = query.replace(/#/g, '').trim();
+    var searchHash = { per_page: 60, page: 0 };
 
-    if (filters === "MY_POSTS" && userData()) {
-      searchObj["tagFilters"] = [['user_' + userData()['id']]]
-    } else if (hashtags != null && hashtags.length > 0) {
+    if (hashtags != null && hashtags.length > 0) {
```
Comment (nickytonline) [inline app/views/articles/_search.html.erb]: This will check if it's `null` or `undefined`. I'm assuming the value would only ever be `null`, `undefined` or an array.

```suggestion
    if (hashtags && hashtags.length > 0) {
```

File: app/views/articles/_search.html.erb [hunk 7]
```diff
@@ -45,68 +43,64 @@
...
       for(var i = 0; i < hashtags.length; i++){
-        hashtags[i] = hashtags[i].replace(/#/,'')
+        hashtags[i] = hashtags[i].replace(/#/,'');
       }
-      searchObj["tagFilters"] = hashtags
-    } else {
-      searchObj["filters"] = filters
+      searchHash["tag_names"] = hashtags;
+    } else if (filters && filters != "") {
```
Comment (nickytonline) [inline app/views/articles/_search.html.erb]: This is fine, but the linter will complain about it not being `!==`

```suggestion
    } else if (filters && filters !== "") {
```

File: app/views/articles/_search.html.erb [hunk 8]
```diff
@@ -45,68 +43,64 @@
...
+      filters.split('&').forEach((filter) => {
+        const [key, value] = filter.split(':');
+        searchHash[key] = value;
+      });
     }
-    index.search(query, searchObj)
+
+    if (searchTerm && searchTerm != "") { searchHash["search_fields"] = searchTerm; }
```
Comment (nickytonline) [inline app/views/articles/_search.html.erb]: Appeasing the linter gods.

```suggestion
    if (searchTerm && searchTerm !== "") { searchHash["search_fields"] = searchTerm; }
```

File: app/javascript/articles/components/SearchSnippet.jsx (hunk 5 above)
Comment (mstruve) [reply app/javascript/articles/components/SearchSnippet.jsx]: body_text is now an array of two strings. Before it was only a single string bc we only took the first highlighted match and presented it. I decided to take two strings to try and give the user a little more insight into what matched for their search.

File: app/views/articles/_search.html.erb (hunk 6 above)
Comment (mstruve) [reply app/views/articles/_search.html.erb]: I would assume so, that code I just left from before but I'm game for an easy refactor.

File: app/views/articles/_search.html.erb [hunk 9]
```diff
@@ -45,68 +43,64 @@
...
+    fetch(`/search/feed_content?${searchParams}`, {
+      method: 'GET',
+      headers: {
+        Accept: 'application/json',
+        'X-CSRF-Token': window.csrfToken,
+        'Content-Type': 'application/json',
+      },
+      credentials: 'same-origin',
+    })
+      .then(response => response.json())
```
Comment (atsmith813) [inline app/views/articles/_search.html.erb]: Can we use the [`fetchSearch` helper function](https://github.com/atsmith813/dev.to/blob/9a0a6fbe8f3b45df600ba3a1571ab8164fd1bc41/app/javascript/src/utils/search.js#L90-L102) here?

```javascript
const responsePromise = fetchSearch('feed_content', searchHash)

responsePromise.then((content) => {
  // continuing on...
}
```

File: app/assets/javascripts/initializers/initScrolling.js.erb [hunk 10]
```diff
@@ -179,40 +179,46 @@ function fetchNextPodcastPage(el) {
...
+  fetch(`/search/feed_content?${searchParams}`, {
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then(response => response.json())
```
Comment (atsmith813) [inline app/assets/javascripts/initializers/initScrolling.js.erb]: Same here 👀 

Can we use the [`fetchSearch` helper function](https://github.com/atsmith813/dev.to/blob/9a0a6fbe8f3b45df600ba3a1571ab8164fd1bc41/app/javascript/src/utils/search.js#L90-L102) here?

```javascript
const responsePromise = fetchSearch('feed_content', searchHash)

responsePromise.then((content) => {
  // continuing on...
}
```

File: app/views/articles/_search.html.erb (hunk 9 above)
Comment (rhymes) [reply app/views/articles/_search.html.erb]: You could even dare to try `async/await` as it's now been enabled:

```js
const result = await fetchSearch(...);
// do something with the result
```

I'll double check to see if it actually works :D

**Comment by mstruve** [comment]
> Why did we move from q to search_fields ?

I did this for 2 reasons. For one, `search_fields` is the field in Elasticsearch that we are searching so rather than having to parse the params in the controller I thought it would be easier to pass the param we need.

The second reason was that when I started testing I kept finding new places I needed to update the code in the javascript. I eventually decided that if I forced myself to use `search_fields` then I would have to go and find every q= and replace it which actually worked really well for finding all the spots we were calling that search endpoint. 

> Other than testing it from the homepage, clicking on the filters is there anything else we might do for manual testing?

Yes! I tested the homepage signed in and signed out. I also tested clicking through to tag pages and filtering down there. Test the scrolling functions on the homepage and tag pages. Test going to a user's page and scrolling down. 

I'll work on those refractors.

File: app/views/articles/_search.html.erb (hunk 9 above)
Comment (mstruve) [reply app/views/articles/_search.html.erb]: @atsmith813 we cant use that helper here or in the other file for the following reasons as explained to me by @rhymes 

> I don’t think it’ll work but better check first. The reason why it doesn’t is that the code you’re trying to refactor by using fetchSearch  is handled by sprockets, but the fetchSearch resides in our webpack code, which is imported only at will and not known by app/assets/javascript . Long story short. app/javascript  can import code from app/assets/javascripts  but not the other way around
the only thing you can do is to move the function inside app/assets/javascripts  and call it from both places but i think it’s outside of the scope of the PR

Comment (atsmith813) [reply app/views/articles/_search.html.erb]: Makes sense. That's kind of funny, that's why I didn't include the import statement in the code example and just wrote _"after importing it"_ because I realized I was unclear on how that would actually be imported 🤔TIL and this is a good example of [our worst technical debt](https://docs.dev.to/technical-overview/architecture/#worst-technical-debt). Thanks for the info, both of you!

**Comment by mstruve** [comment]
@rhymes I switched us back to using `q=` on the frontend bc @benhalpern mentioned its kind of the standard and what people expect.

**Comment by nickytonline** [comment]
Another thing I noticed @mstruve is if I search for a word that is in several article titles, only one hit is returned for posts (in my example at least).

image

If I go to the comments search section, I see multiple posts with that word in the title and one duplicate article and this word is not included in any of the comments.

image

image

**Comment by mstruve** [comment]
From my testing podcast filtering works well as well as finding words in titles. 

Screen Shot 2020-03-31 at 11 15 50 AM

Screen Shot 2020-03-31 at 11 16 00 AM

A likely cause of stuff not showing up is Elasticsearch not being synced for some reason with Postgres. If the data is in Elasticsearch, it should be showing up. I have also found that a hard shutdown and startup is helpful. It also sometimes takes a few refreshes at the beginning before all the javascript works correctly for me.

**Comment by benhalpern** [comment]
@rhymes we implemented a minimum `score` of 3 in the "feed" portion of a tag, with content otherwise being relegated to `/latest`. Since your local content has a default of 0 I think that's what's going on. This doesn't seem related to the work in this pr.

**Comment by rhymes** [comment]
> @rhymes we implemented a minimum `score` of 3 in the "feed" portion of a tag, with content otherwise being relegated to `/latest`. Since your local content has a default of 0 I think that's what's going on. This doesn't seem related to the work in this pr.

Yep, still working! Updated the score to 4 and there they are. Thanks!

File: app/views/articles/_search.html.erb [hunk 11]
```diff
@@ -53,60 +50,58 @@
...
+    if (filters && filters !== "") {
+      filters.split('&').forEach((filter) => {
+        const [key, value] = filter.split(':');
+        searchHash[key] = value;
+      });
     }
-    index.search(query, searchObj)
-      .then(function searchDone(content) {
+
+    if (searchTerm && searchTerm !== "") { searchHash.search_fields = searchTerm; }
```
Comment (rhymes) [inline app/views/articles/_search.html.erb]: same here

File: app/views/articles/index.html.erb [hunk 12]
```diff
@@ -30,7 +30,7 @@
       data-params="<%= params.to_json(only: %i[tag username q]) %>" data-which="<%= @list_of %>"
       data-tag=""
       data-feed="<%= params[:timeframe] || "base-feed" %>"
-      data-articles-since="<%= Timeframer.new(params[:timeframe]).datetime.to_i %>">
+      data-articles-since="<%= Timeframer.new(params[:timeframe]).datetime&.iso8601 %>">
```
Comment (rhymes) [inline app/views/articles/index.html.erb]: where does this get parsed? I was trying to find it in the code, to make sure that it wouldn't break with empty string but I can't find it in the JS

Comment (mstruve) [reply app/views/articles/index.html.erb]: https://github.com/thepracticaldev/dev.to/pull/6847/files#diff-0869f8ed3a2fce756f2dc6a68000ed74R199

If it is a blank value we will remove it from the params in the controller.

File: app/javascript/articles/components/SearchSnippet.jsx [hunk 13]
```diff
@@ -1,45 +1,25 @@
...
-export const SearchSnippet = ({ snippetResult }) => {
-  if (snippetResult && snippetResult.body_text) {
+export const SearchSnippet = ({ highlightText }) => {
+  if (highlightText && highlightText.body_text.length > 0) {
+    const hitHighlights = highlightText.body_text;
     let bodyTextSnippet = '';
 
-    if (snippetResult.body_text.matchLevel !== 'none') {
-      const firstSnippetChar = snippetResult.body_text.value[0];
+    if (hitHighlights[0] && hitHighlights[0] !== '') {
```
[the remark this answers is not in the document]
Comment (mstruve) [reply app/javascript/articles/components/SearchSnippet.jsx]: TIL! Thank you!!!
